ambry-store/src/main/java/com.github.ambry.store/CompactionLog.java

```diff
@@ -0,0 +1,336 @@
+      CrcInputStream crcInputStream = new CrcInputStream(fileInputStream);
+      DataInputStream stream = new DataInputStream(crcInputStream);
+      short version = stream.readShort();
+      switch (version) {
+        case VERSION_0:
+          startTime = stream.readLong();
+          currentIdx = stream.readInt();
+          int cycleLogsSize = stream.readInt();
+          cycleLogs = new ArrayList<>(cycleLogsSize);
+          while (cycleLogs.size() < cycleLogsSize) {
+            cycleLogs.add(CycleLog.fromBytes(stream, storeKeyFactory));
+          }
+          long crc = crcInputStream.getValue();
+          if (crc != stream.readLong()) {
+            throw new IllegalStateException("CRC of data read does not match CRC in file");
+          }
+          break;
+        default:
+          throw new IllegalArgumentException("Unrecognized version");
+      }
+    }
+  }
+
+  /**
+   * @return the current state of compaction.
+   */
+  State getCompactionState() {
+    return currentIdx >= cycleLogs.size() ? State.DONE : getCurrentCycleLog().getState();
+  }
+
+  /**
+   * @return the {@link CompactionDetails} for the compaction cycle in progress.
+   */
+  CompactionDetails getCompactionDetails() {
+    return getCurrentCycleLog().compactionDetails;
+  }
+
+  /**
+   * @return the {@link StoreFindToken} until which data has been copied and flushed.
+   */
+  StoreFindToken getSafeToken() {
+    return getCurrentCycleLog().safeToken;
+  }
+
+  /**
+   * Sets the {@link StoreFindToken} until which data is copied and flushed.
+   * @param safeToken the {@link StoreFindToken} until which data is copied and flushed.
+   */
+  void setSafeToken(StoreFindToken safeToken) {
+    CycleLog cycleLog = getCurrentCycleLog();
+    if (!cycleLog.getState().equals(State.COPY)) {
+      throw new IllegalStateException("Cannot set a safe token - not in COPY phase");
+    }
+    cycleLog.safeToken = safeToken;
+    flush();
+  }
+
+  /**
+   * Marks the start of the copy phase.
+   */
+  void markCopyStart() {
+    getCurrentCycleLog().copyStartTime = time.milliseconds();
+    flush();
+  }
+
+  /**
+   * Marks the start of the switch phase.
+   */
+  void markSwitchStart() {
+    getCurrentCycleLog().switchStartTime = time.milliseconds();
+    flush();
+  }
+
+  /**
+   * Marks the start of the cleanup phase.
+   */
+  void markCleanupStart() {
+    getCurrentCycleLog().cleanupStartTime = time.milliseconds();
+    flush();
+  }
+
+  /**
+   * Marks the current compaction cycle as complete.
+   */
+  void markCycleComplete() {
+    getCurrentCycleLog().cycleEndTime = time.milliseconds();
+    currentIdx++;
+    flush();
+  }
+
+  /**
+   * Closes the compaction log. If compaction is complete, renames the log to keep a permanent record.
+   */
+  @Override
+  public void close() {
+    if (file.exists() && getCompactionState().equals(State.DONE)) {
+      File savedLog = new File(file.getAbsolutePath() + "_" + startTime);
```

Comment:
@asinghai brought up a point that if this is for debugging, it is possible that the file is moved to different servers. In the light of that, I'll see if this can be changed.